docs/v1/P4Runtime-Spec.mdk

```diff
@@ -3942,18 +4037,26 @@ In order to support the above SDN use case, P4Runtime requires translation of
 port metadata values between the controller's space and the PSA device's space
 as needed. Such translation is enabled by identifying a P4 entity (match field,
 action parameter or header field) as being a PSA port metadata type. For this
-purpose, PSA defines the port metadata field type using special P4 types, namely
-`PortId_t` and `PortInHeader_t`, instead of standard P4 bitstrings. The P4Info
-for all P4 entities of the special PSA port types use a controller-defined
-32-bit type instead of the dataplane bitwidth defined in the P4 program. The
-following PSA port metadata types are defined in *psa.p4* for the PSA device in
-Switch 1.
+purpose, PSA defines the port metadata field type using special [user-defined P4
+types](#sec-user-defined-types), namely `PortId_t` and `PortInHeader_t`, instead
+of standard P4 bitstrings. The P4Info for all P4 entities of the special PSA
+port types use a controller-defined 32-bit type instead of the dataplane
+bitwidth defined in the P4 program. The following PSA port metadata types are
+defined in *psa.p4* for the PSA device in Switch 1.
 
 ~ Begin P4Example
+@p4runtime_translation("p4.org/psa/v1/PortId_t", 32)
 type PortId_t bit<9>;
+@p4runtime_translation("p4.org/psa/v1/PortIdInHeader_t", 32)
 type PortIdInHeader_t bit<32>;
 ~ End P4Example
 
+The first argument to the `@p4runtime_translation` annotation is a URI that
+indicates to the P4Runtime server which numerical mapping - provided by the
+out-of-band switch configuration mechanism - to use to translate between the SDN
+value and the dataplane value. The second argument is the bitwidth of the SDN
+representation of the transltaed entity (32-bit in the case of ports).
+
```

Comment:
Do we have a separate open issue to track the comment below of "PSA device port number 0 ... invalid"?  I know your brought this up at a P4 API work group meeting.  I believe the _data plane_ should allow ingress and egress port number 0 to be valid, even if the P4Runtime messages use numeric translation so that 0 is considered invalid there.

Reply:
Yes, we have an issue, you even commented there :)
https://github.com/p4lang/p4runtime/issues/49

Reply:
They say that memory is the second thing to go :)